File: 프로그래머스/lv2/42577. 전화번호 목록/전화번호 목록.py
def solution(phone_book):
    answer = True
    
    phone_dict ={}
    
    for i, phone in enumerate(phone_book):
        phone_dict[phone] = i
    
    for phone in phone_book:
        tmp = ''
        for p in phone:
            tmp+=p
            if tmp in phone_dict.keys() and phone != tmp:
                answer = False
                break
    return answer

File: 프로그래머스/2/42577. 전화번호 목록/전화번호 목록.java
import java.util.*;

class Solution {
    public boolean solution(String[] phone_book) {
        boolean answer = true;
        
        HashSet<String> set = new HashSet<>();
        
        for (String phoneNum : phone_book){
            set.add(phoneNum);
        }
        
        for (int i =0; i< phone_book.length;i++){
            for (int j = 0; j < phone_book[i].length(); j++) {
                String curPrefix = phone_book[i].substring(0, j);
                if (set.contains(curPrefix)) return false;
            }
        }
        
        return answer;
    }
}